notebooks/tutorial_ivf_pq.ipynb

```diff
@@ -0,0 +1,1226 @@
+    "        n += cp.intersect1d(found_indices[i, :k], ground_truth[i, :k]).size\n",
+    "    recall = n / found_indices.size\n",
+    "    return recall\n",
+    "\n",
+    "recall_first_try = calc_recall(neighbors, gt_neighbors)\n",
+    "print(f\"Got recall = {recall_first_try} with the default parameters (k = {k}).\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Refine\n",
+    "Let's improve our results a little bit!\n",
+    "The refinement operation follows an approximate NN search.\n",
+    "It recomputes the exact distances for the already selected candidates and selects a subset of them thus improving the recall."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from pylibraft.neighbors import refine"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "CPU times: user 136 ms, sys: 7.79 ms, total: 143 ms\n",
+      "Wall time: 143 ms\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%time\n",
+    "\n",
+    "candidates = ivf_pq.search(search_params, index, queries, k * 2, handle=resources)[1]\n",
+    "distances, neighbors = refine(dataset, queries, candidates, k, handle=resources)\n",
+    "resources.sync()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Got recall = 0.94815 with 2x refinement (k = 10).\n"
+     ]
+    }
+   ],
+   "source": [
+    "recall_refine2x = calc_recall(neighbors, gt_neighbors)\n",
+    "print(f\"Got recall = {recall_refine2x} with 2x refinement (k = {k}).\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Tweaking search parameters"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Number of neighbors\n",
+    "Let see how QPS depens on `k`. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "12.8 ms ± 1.51 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)\n",
+      "13.6 ms ± 1.73 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)\n",
+      "13.8 ms ± 926 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)\n",
+      "14.1 ms ± 1.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)\n",
+      "14.8 ms ± 1.41 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)\n",
+      "18.1 ms ± 3.24 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)\n",
+      "19.9 ms ± 13.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)\n",
+      "30.6 ms ± 37.5 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)\n",
```

Comment:
Why this drop at k=128, is it an sisue of top-k kernel selection?

Reply:
There seems to be a performance regression, which results in IVF-PQ in the range of `k = 65-128` being slower than for `k = 129...~400`. I observed this in internal benchmarks, and probably it's also manifested in https://github.com/rapidsai/raft/issues/1621